File: 1701-average-waiting-time/1701-average-waiting-time.dart
import 'dart:math';
class Solution {
  double averageWaitingTime(List<List<int>> customers) {
    int finTime = 0;
    int res = 0;
    for (final cus in customers){
      finTime = max(finTime, cus[0]) + cus[1];
      res += finTime - cus[0];
    }
    return res / customers.length;
  }
}

File: 1701-average-waiting-time/1701-average-waiting-time.ex
defmodule Solution do
  @spec average_waiting_time(customers :: [[integer]]) :: float
  def average_waiting_time(customers) do
    customers
    |> Enum.reduce({0, 0}, &sum_and_wait/2)
    |> (fn {_, wait} -> wait / length(customers) end).()
  end
  defp sum_and_wait([arv, t], {sum, wait}) do
    pre_sum = max(sum, arv)
    {pre_sum + t, pre_sum + wait + t - arv}
  end
end